implement NumPy-like functionality maximum, minimum

RockingJavaBean opened the pull request:
Related to #38349

Implement NumPy-like functions `maximum` and `minimum`.
The `maximum` and `minimum` functions compute input tensors element-wise, returning a new array with the element-wise maxima/minima.

If one of the elements being compared is a NaN, then that element is returned, both `maximum` and `minimum` functions do not support complex inputs.

This PR also promotes the overloaded versions of torch.max and torch.min, by re-dispatching binary `torch.max` and `torch.min` to `torch.maximum` and `torch.minimum`.

RockingJavaBean committed:
define native function

RockingJavaBean committed:
implement maximum

RockingJavaBean committed:
implement minimum

RockingJavaBean committed:
add tests

RockingJavaBean committed:
add docs

RockingJavaBean committed:
fix macos build issue

RockingJavaBean commented:
@mruberry This PR implements a pair of NumPy functions `maximum` and `minimum`, please kindly help review.

RockingJavaBean commented:
It seems `pr/pytorch-linux-xenial-rocm3.5.1-py3.6` failed due to the build timeout.

```
00:58:39 Build timed out (after 180 minutes). Marking the build as failed.
00:58:39 /tmp/jenkins2758323857757916016.sh: line 208: 75959 Done                    ( echo "cd $docker_workspace"; if [ "$IMPORT_ENV" == 1 ]; then
00:58:39     echo 'source ./env';
00:58:39 fi; echo 'declare -x WORKSPACE=$PWD'; sed -n '/^####/ { s///; :a; n; p; ba; }' "${BASH_SOURCE[0]}" )
00:58:39      75960 Terminated              | docker exec $docker_user -i "$id" bash
00:58:39 + echo 'Stopping container...'
00:58:39 Stopping container...
```

mruberry commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -789,6 +791,30 @@ Tensor min(const Tensor& self, const Tensor& other) {
+}
+
+Tensor maximum(const Tensor& self, const Tensor& other) {
+  Tensor result = at::empty(0, self.options());
```
`Tensor result;`

RockingJavaBean replied:
Thanks for pointing this out, I have changed the code to `Tensor result` and used `iter.output()` for the returning value.

mruberry commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -789,6 +791,30 @@ Tensor min(const Tensor& self, const Tensor& other) {
+}
+
+Tensor minimum(const Tensor& self, const Tensor& other) {
+  Tensor result = at::empty(0, self.options());
```
`Tensor result;`

RockingJavaBean replied:
Addressed.

mruberry commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -502,6 +502,80 @@ void min_elementwise_kernel(TensorIterator& iter) {
+  } else {
+    AT_DISPATCH_COMPLEX_TYPES(iter.dtype(), "maximum_cpu", [&] {
+      cpu_kernel(iter, [](scalar_t a, scalar_t b) -> scalar_t {
+        if (_isnan(a.real()) || _isnan(a.imag())) {
```
I think isnan works on complex, see:

https://github.com/pytorch/pytorch/blob/2f840b1662b487d5551d7230f8eb4d57645cfff5/aten/src/ATen/native/TensorCompare.cpp#L80

It's documented as doing so:

https://pytorch.org/docs/master/generated/torch.isnan.html?highlight=isnan#torch.isnan

This should save you from looking at the real and imaginary parts separately.

mruberry replied:
However... (see above)

RockingJavaBean replied:
Thanks for the detailed explanation of `isnan` function.
As complex numbers are not supported, the complex type dispatch codes are removed in the latest commit.

mruberry commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -502,6 +502,80 @@ void min_elementwise_kernel(TensorIterator& iter) {
+      });
+    });
+  } else {
+    AT_DISPATCH_COMPLEX_TYPES(iter.dtype(), "maximum_cpu", [&] {
```
Comparisons between complex numbers aren't well-defined mathematically and we prohibit them elsewhere (NumPy is updating their functionality to conform to ours). I would think that means that we shouldn't support maximum and minimum for complex tensors?

RockingJavaBean replied:
Thanks for the kind suggestions, I have added checks to prevent complex inputs and deleted redundant complex dispatch codes.

mruberry commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -502,6 +502,80 @@ void min_elementwise_kernel(TensorIterator& iter) {
+  }
+}
+
+void minimum_kernel(TensorIterator& iter) {
```
See symmetric comments above.

RockingJavaBean replied:
addressed

RockingJavaBean replied:
addressed

mruberry commented on aten/src/ATen/native/cuda/MaximumMinimum.cu:
```diff
@@ -0,0 +1,94 @@
+
+namespace at { namespace native {
+
+void maximum_kernel_cuda(TensorIterator& iter) {
```
See symmetric comments above.

RockingJavaBean replied:
addressed

mruberry commented on aten/src/ATen/native/cuda/MaximumMinimum.cu:
```diff
@@ -0,0 +1,94 @@
+    AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.input_dtype(), "maximum_cuda", [&]() {
+      gpu_kernel(iter, [] GPU_LAMBDA (scalar_t a, scalar_t b) -> scalar_t {
+        // isnan(half) breaks the Windows build. We explicitly cast half to float.
+        using acc_type = typename AccumulateType<scalar_t, /*is_cuda=*/true>::type;
```
Why use the accumulate type here?

RockingJavaBean replied:
I was referring the following code from `MaxMinElementwiseKernel.cu` to prevent `isnan` breaks for `half` types on the Windows platform.
https://github.com/pytorch/pytorch/blob/e8f4b04d9ae1880fd8d09dde40639e86a0d2c585/aten/src/ATen/native/cuda/MaxMinElementwiseKernel.cu#L26-L32

And thanks to the kind suggestions on using `a != a` directly, the `accumulate type` is no longer needed.

mruberry commented on aten/src/ATen/native/cuda/MaximumMinimum.cu:
```diff
@@ -0,0 +1,94 @@
+  } else if (isFloatingType(iter.dtype())) {
+    AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.input_dtype(), "minimum_cuda", [&]() {
+      gpu_kernel(iter, [] GPU_LAMBDA (scalar_t a, scalar_t b) -> scalar_t {
+        // isnan(half) breaks the Windows build. We explicitly cast half to float.
```
Would `(a != a)` work?

RockingJavaBean replied:
Yes, it works and I have used `a != a` directly for checking nans.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5197,6 +5197,22 @@
+  variants: method, function
+
+- func: maximum.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
+  dispatch:
```
I don't think you need to be explicit about this dispatch. I believe it will automatically dispatch to `maximum_out`.

RockingJavaBean replied:
Thanks for pointing that out, I have removed the redundant dispatch sections.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5197,6 +5197,22 @@
+  variants: method, function
+
+- func: minimum.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
+  dispatch:
```
Same note on dispatch here.

RockingJavaBean replied:
addressed.

mruberry commented on test/test_torch.py:
```diff
@@ -15389,6 +15389,88 @@ def test_minmax(self, device, dtype):
 
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    @dtypes(*(torch.testing.get_all_int_dtypes() + [torch.bool]))
+    def test_maximum_minimum_int_and_bool(self, device, dtype):
```
This test is already very good. You could improve it by generating the NumPy arrays using https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.integers.html?highlight=integers#numpy.random.Generator.integers instead of handwriting a tuple of values. The arrays can then be converted to torch tensors with `torch.from_numpy(a).to(device=device)`.

RockingJavaBean replied:
Thanks for the kind suggestions, I have updated `test_maximum_minimum_int_and_bool` to let NumPy generate random inputs for me.

mruberry commented on test/test_torch.py:
```diff
@@ -15389,6 +15389,88 @@ def test_minmax(self, device, dtype):
+            b_np = np.array(b_vals, dtype=torch_to_numpy_dtype_dict[dtype])
+            numpy_result = numpy_op(a_np, b_np)
+
+            self.assertEqual(tensor_result.cpu(), torch.from_numpy(numpy_result))
```
I think you can compare directly to the NumPy array without converting it to a `torch.Tensor`.

RockingJavaBean replied:
Addressed.

mruberry commented on test/test_torch.py:
```diff
@@ -15389,6 +15389,88 @@ def test_minmax(self, device, dtype):
+        # np.maximum and np.minimum functions compare input arrays element-wisely.
+        # if one of the elements being compared is a NaN, then that element is returned.
+        ops = ((torch.maximum, np.maximum), (torch.minimum, np.minimum))
+        a_vals = (float('inf'), -float('inf'), 2.0, float('nan'), -1.0, float('nan'))
```
You could also simplify this test using `np.randn` to generate a float tensor, if you like. You would still have to generate an "extremal values" test case with -inf, inf, and nan by hand, though. You also can simplify by just comparing the torch and numpy results. You don't need to assert the results are correct. 

To handle bfloat16 you can just cast the generated arrays to the desired torch type (leaving the arrays in double) and then compare using `self.assertEqual(actual, expected, exact_dtype=False)`.

RockingJavaBean replied:
Thanks for the suggestions, I have split the test cases into two methods
- test_maximum_minimum_float: use `np.randn` to generate random float inputs.
- test_maximum_minimum_float_nan_and_inf: test external values like `nan`, `-inf` and `inf`

For `bfloat16`,  I have to set rtol and atol like below to prevent precision issues 
```
self.assertEqual(tensor_result, numpy_result, rtol=0.016, atol=1e-5, exact_dtype=False)
````
The values of rtol and atol refers to `common_utils.py` file.
https://github.com/pytorch/pytorch/blob/42b4a7132e7c6f1df963b473d1583e4791fb1808/torch/testing/_internal/common_utils.py#L939-L942

mruberry commented on test/test_torch.py:
```diff
@@ -15389,6 +15389,88 @@ def test_minmax(self, device, dtype):
+
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    @dtypes(torch.complex64, torch.complex128)
+    def test_maximum_minimum_complex(self, device, dtype):
```
See above comments: should we support complex?

RockingJavaBean replied:
The test cases for complex numbers have been updated.
The current test cases are to whether the error messages are raised when either input is a complex number.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3860,6 +3860,34 @@ def merge_dicts(*dicts):
+
+.. note::
+    If one of the elements being compared is a NaN, then that element is returned.
+    If both elements are NaNs then the first is returned.
```
This part of the note may be removed.

RockingJavaBean replied:
The `If both elements are NaNs then the first is returned.` is removed and the whole note is changed like `minimum` function.
Please kindly confirm.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3860,6 +3860,34 @@ def merge_dicts(*dicts):
+add_docstr(torch.maximum, r"""
+maximum(input, other, *, out=None) -> Tensor
+
+Compares the element-wise maximum of :attr:`input` and :attr:`other`.
```
"Computes the..."

RockingJavaBean replied:
Addressed

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3860,6 +3860,34 @@ def merge_dicts(*dicts):
+    >>> torch.maximum(torch.tensor([[2.1, 1], [float('inf'), 6]]), torch.tensor([[1.2, 2], [4, float('nan')]]))
+    tensor([[2.1000, 2.0000],
+            [   inf,    nan]])
+    >>> torch.maximum(torch.tensor([complex(2, 1), complex(1, 2)]), torch.tensor([complex(1, 10), complex(1, 8)]))
```
These examples may be removed.

RockingJavaBean replied:
The examples of complex numbers are removed.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -4102,6 +4130,34 @@ def merge_dicts(*dicts):
+add_docstr(torch.minimum, r"""
+minimum(input, other, *, out=None) -> Tensor
+
+Compares the element-wise minimum of :attr:`input` and :attr:`other`.
```
"Computes the..."

RockingJavaBean replied:
Addressed

mruberry commented on torch/_torch_docs.py:
```diff
@@ -4102,6 +4130,34 @@ def merge_dicts(*dicts):
+
+.. note::
+    If one of the elements being compared is a NaN, then that element is returned.
+    If both elements are NaNs then the first is returned.
```
This note may need updating.

RockingJavaBean replied:
I have changed the notes to below
```
.. note::
    If one of the elements being compared is a NaN, then that element is returned.
    :func:`minimum` is not supported for tensors with complex dtypes.
```

mruberry commented on torch/_torch_docs.py:
```diff
@@ -4102,6 +4130,34 @@ def merge_dicts(*dicts):
+    >>> torch.minimum(torch.tensor([[2.1, 1], [-float('inf'), 6]]), torch.tensor([[1.2, 2], [4, float('nan')]]))
+    tensor([[1.2000, 1.0000],
+            [  -inf,    nan]])
+    >>> torch.minimum(torch.tensor([complex(2, 1), complex(1, 2)]), torch.tensor([complex(1, 10), complex(1, 8)]))
```
These examples may be removed.

RockingJavaBean replied:
The examples of complex numbers are removed.

mruberry reviewed (changes requested):
Thanks for the PR, @RockingJavaBean! This is very well written. I made some comments/suggestions. Most notably, should these functions support complex tensors? We typically don't allow comparisons like a > b for complex numbers because they're not part of any ordered field. 

I also want to ask about something not in this PR: autograd. @albanD's out this week but I'd like him to suggest how to deal with autograd for maximum/minimum. I would think our treatment would be analogous to `torch.where`? Let's wait for him to tell us.

RockingJavaBean committed:
update for feedbacks

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into numpy_maximum_minimum_bk

RockingJavaBean committed:
update docs and testcases

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into numpy_maximum_minimum

RockingJavaBean committed:
fix xla error by adding dispatch sections

RockingJavaBean committed:
autograd for maximum and minimum

RockingJavaBean commented on tools/autograd/derivatives.yaml:
```diff
@@ -1082,6 +1082,16 @@
+- name: maximum(Tensor self, Tensor other) -> Tensor
+  condition: non_differentiable
+  self: where((self != self) | ((other == other) & (self >= other)), grad, zeros_like(grad, at::MemoryFormat::Preserve))
+  other: where((self != self) | ((other == other) & (self >= other)), zeros_like(grad, at::MemoryFormat::Preserve), grad)
```
The derivatives for `maximum` and `minimum` are implemented according to the `torch.where` analogy.
The reason for using `self != self` is to check `nan`.
If one of the elements being compared is a NaN, then that element is returned by `maximum` and `minimum`  functions.

RockingJavaBean commented:
@mruberry thank you so much for the kind suggestions and throughout review for this PR.
I updated the code to prevent complex inputs for both `maximum` and `minimum` functions.
Please kindly review.

RockingJavaBean commented:
@mruberry @albanD 
I tried to implement the autograd for maximum and minimum functions based on the `torch.where` analogy via 4d997e1.
Below are the comparisons between `torch.where` with `maximum/minimum`.

```python
>>> import torch
>>> x = torch.tensor([1., 2., 3.], requires_grad=True)
>>> y = torch.tensor([2., 1., 5.], requires_grad=True)
>>> z = torch.where(x > y, x, y)
>>> z.backward(torch.tensor([1., 1., 1.]))
>>> x.grad
tensor([0., 1., 0.])
>>> y.grad
tensor([1., 0., 1.])

>>> x = torch.tensor([1., 2., 3.], requires_grad=True)
>>> y = torch.tensor([2., 1., 5.], requires_grad=True)
>>> z = torch.maximum(x, y)
>>> z.backward(torch.tensor([1., 1., 1.]))
>>> x.grad
tensor([0., 1., 0.])
>>> y.grad
tensor([1., 0., 1.])

>>> x = torch.tensor([1., 2., 3.], requires_grad=True)
>>> y = torch.tensor([2., 1., 5.], requires_grad=True)
>>> z = torch.minimum(x, y)
>>> z.backward(torch.tensor([1., 1., 1.]))
>>> x.grad
tensor([1., 0., 1.])
>>> y.grad
tensor([0., 1., 0.])
```

I think they match the patterns.

mruberry commented:
Awesome! Thanks for the update @RockingJavaBean. 

@albanD actually pointed out that `torch.max` and `torch.min` have similar functionality! Which is pretty well hidden. We'd like to promote `torch.maximum` and `torch.minimum` over those hidden overloaded versions of `torch.max` and `torch.min`. Would you mind if I update this PR myself with a few edits to clarify that people should be using `torch.maximum` and `torch.minimum` where possible?

RockingJavaBean commented:
@mruberry No, certainly not, please kindly edit this pull request.

RockingJavaBean commented:
@mruberry thanks so much for the guidance throughout this PR, and your help on updating with a few edits to clarify that people should be using `maximum/minimum` where possible.

Please kindly let me know if there is anything I can do for help.

mruberry commented:
> @mruberry thanks so much for the guidance throughout this PR, and your help on updating with a few edits to clarify that people should be using `maximum/minimum` where possible.
> 
> Please kindly let me know if there is anything I can do for help.

Thank you, @RockingJavaBean. I'll be able to take a look and maybe make a few tweaks later today.

Commit:
Merge branch 'master' of https://github.com/pytorch/pytorch into numpy_maximum_minimum

Commit:
Prioritizes minimum, maximum over binary min, max

mruberry commented:
@RockingJavaBean please take a look.

I merged your kernels with the existing binary min and max kernels and updated the documentation to refer to minimum and maximum. I also added a few tests, including one for type promotion.

RockingJavaBean commented:
> I merged your kernels with the existing binary min and max kernels and updated the documentation to refer to minimum and maximum. I also added a few tests, including one for type promotion.

@mruberry Thank you so much for helping to update this PR.
I have looked into your commits and find it very inspiring for me.

The only question I have is about `autograd`.
```yaml
- name: maximum(Tensor self, Tensor other) -> Tensor
  self: grad.clone().masked_fill_(self <= other, 0)
  other: grad.clone().masked_fill_(self > other, 0)
```

I think the above implementations do not handle the `nan`.
Please kindly confirm whether it is feasible to ignore `nan` in terms of the autograd of `maximum/minimum`, or should we add associated backward functions in `tools/autograd/templates/Functions.cpp`?

mruberry commented:
> > I merged your kernels with the existing binary min and max kernels and updated the documentation to refer to minimum and maximum. I also added a few tests, including one for type promotion.
> 
> @mruberry Thank you so much for helping to update this PR.
> I have looked into your commits and find it very inspiring for me.
> 
> The only question I have is about `autograd`.
> 
> ```yaml
> - name: maximum(Tensor self, Tensor other) -> Tensor
>   self: grad.clone().masked_fill_(self <= other, 0)
>   other: grad.clone().masked_fill_(self > other, 0)
> ```
> 
> I think the above implementations do not handle the `nan`.
> Please kindly confirm whether it is feasible to ignore `nan` in terms of the autograd of `maximum/minimum`, or should we add associated backward functions in `tools/autograd/templates/Functions.cpp`?

That's a great question. Luckily, we get to ask @albanD for his thoughts.

@albanD, for binary max the gradient formula was what you see above (I changed the name to maximum, don't worry about it). @RockingJavaBean developed the following, which is interesting:

```
- name: maximum(Tensor self, Tensor other) -> Tensor
  condition: non_differentiable
  self: where((self != self) | ((other == other) & (self >= other)), grad, zeros_like(grad, at::MemoryFormat::Preserve))
  other: where((self != self) | ((other == other) & (self >= other)), zeros_like(grad, at::MemoryFormat::Preserve), grad)
```

Would you like to weigh-in, as our autograd expert?

albanD commented:
Hi,

I would have three questions:
- How does the old max handle nans?  How is it different from maximum?
- In which cases the old formula was not correct?
- What is the perf impact of this change? It seems that we go from going 3 times through something of the size of the input to  6 (if I count correctly). Which might be problematic.

mruberry commented:
> Hi,
> 
> I would have three questions:
> 
> * How does the old max handle nans?  How is it different from maximum?
> * In which cases the old formula was not correct?
> * What is the perf impact of this change? It seems that we go from going 3 times through something of the size of the input to  6 (if I count correctly). Which might be problematic.

In the interest of PR throughput, I think we'd both agree that we can continue to use the historic grad formula for now, and revisit changing it in a separate PR later if we want to, right?

albanD commented:
If the forward behavior did not change, then no need to change the backward behavior indeed.
But we might need to double check that formula indeed in a follow up PR to make sure it works properly when the inputs has nan.

mruberry commented:
@RockingJavaBean does separating the potential autograd updates to a separate PR sound OK with you? How does the rest of the PR look?

RockingJavaBean commented:
@mruberry remaining historic grad formula in this PR and updating autograd for maximum and minimum functions in a follow-up PR sound great.

The rest of this PR, especially your works on dispatching binary `torch.max` and `torch.min` to `torch.maximum` and `torch.minimum`, looks really good to me.

mruberry reviewed (approved):
Nice teamwork!

mruberry commented:
Just waiting on resolution of https://github.com/pytorch/xla/issues/2424 now.

JackCaoG commented:
Hi @RockingJavaBean , is it possible to rebase this branch with the pytorch master? I submitted https://github.com/pytorch/xla/pull/2438 to resolve the xla issue and pin to this pr as the pytorch version. However, https://github.com/pytorch/pytorch/pull/43047 is missing from this branch and the xla_build will fail. I patched this pr and tested my change locally and everything seems to be fine, but it is still better to let circleCi to do its job.

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into numpy_maximum_minimum

RockingJavaBean commented:
@JackCaoG thanks for helping out with the XLA issue.
I have rebased with the PyTorch master, and now the commit of #43047 is included in this branch.
Please kindly let Circle CI retest the xla_build.

JackCaoG commented:
@RockingJavaBean Thanks! I saw all test passed on XLA:CPU but some failed on XLA:GPU, this is most likely due to a different handling of nan in XLA:GPU. I will check those tomorrow.

mruberry commented:
Thanks for helping out, @JackCaoG!

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into numpy_maximum_minimum

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into numpy_maximum_minimum

RockingJavaBean committed:
remove type promotion test between half and bfloat16

RockingJavaBean commented:
@mruberry The test for type promotion between `half` and `bfloat16` is removed via 965b236 as it has been implemented in #43324.
Please kindly confirm and help land this PR. 
It may also help @JackCaoG lands the related xla PR https://github.com/pytorch/xla/issues/2424.

mruberry commented:
Great work all! Especially @ailzhang  and @JackCaoG for taking the time to help us out with PyTorch/XLA. 

@RockingJavaBean, let me know if you'd like some suggestions for other PyTorch issues or features.

RockingJavaBean commented:
@mruberry I’m so grateful for your kind help and patient guidance throughout this PR.
I love to have your suggestions for other issues or features and continue to make contributions to the PyTorch.

mruberry commented:
> @mruberry I’m so grateful for your kind help and patient guidance throughout this PR.
> I love to have your suggestions for other issues or features and continue to make contributions to the PyTorch.

Happy to help you find another project, @RockingJavaBean. Would you like something simpler or more challenging, another function like maximum and minimum or something different, like a function that used existing PyTorch functions to manipulate tensors, or a function with more challenging numeric and/or performance issues?

RockingJavaBean commented:
@mruberry Thanks for the suggestions.

RockingJavaBean commented:
@mruberry thank you so much for the advice, I would like to try more challenging issues.
Please suggest the `numeric` or `performance` issues that I can start to work on.

mruberry commented:
> @mruberry thank you so much for the advice, I would like to try more challenging issues.
> Please suggest the `numeric` or `performance` issues that I can start to work on.

OK, let me suggest a project and think about whether this would be interesting.

In PyTorch we have torch.linalg.outer, which is an alias for [torch.ger](https://pytorch.org/docs/master/generated/torch.ger.html?highlight=ger#torch.ger), a function which computes the [outer product](https://en.wikipedia.org/wiki/Outer_product) of two vectors. The CPU implementation for this is very old. So old, in fact, that it's in the TH tensor library (see https://github.com/pytorch/pytorch/blob/f06d3904f2c7b6e47ffa532f4a22b9e18941bb90/aten/src/TH/generic/THBlas.cpp#L114), not the more modern ATen tensor library. 

There are several challenges with this function:

- removing the TH code and moving the function's implementation to ATen
- improving the test coverage
- optimizing its performance
- extending it to complex dtypes

These could even be done in 2-3 PRs (handling complex dtypes should definitely be its own PR, the other issues could be combined). 

This would require benchmarking the current BLAS-based solution with a new alternative (like reusing matmul) on various sizes and dtypes, as well as understanding the current code while writing a new, possibly faster implementation.

I think this is a really interesting project, but if it's not what you had in mind let me know and I can suggest something else.  The goal here would be to demonstrate a faster implementation for PyTorch's CPU outer product and modernize the code. Follow-up work could include handling complex vectors or reviewing the CUDA outer product implementation.

RockingJavaBean commented:
@mruberry  thanks so much for the detailed suggestions, it sounds really interesting to me.
I am going to read the [TH-to-ATen-porting-guide](https://github.com/pytorch/pytorch/wiki/TH-to-ATen-porting-guide) for porting the `outer project` first.
Then I would try to improve the test coverage and performance, as well as handling the complex types in the follow-up PRs.

mruberry commented:
> @mruberry thanks so much for the detailed suggestions, it sounds really interesting to me.
> I am going to read the [TH-to-ATen-porting-guide](https://github.com/pytorch/pytorch/wiki/TH-to-ATen-porting-guide) for porting the `outer product` first.
> Then I would try to improve the test coverage and performance, as well as handling the complex types in the follow-up PRs.

Sounds good. Please let me know if you need more pointers or additional help.
